### PR TITLE
Add black, mypy, and pyupgrade to pre-commit

### PR DESCRIPTION
**Pull request by cclauss**
Subtask of  #5843

% `pre-commit run --all-files`
```
make lint-diff...........................................................Passed
Check Yaml...............................................................Passed
black....................................................................Passed
codespell................................................................Passed
mypy.....................................................................Passed
pyupgrade................................................................Passed
```
Add black, mypy, and pyupgrade to pre-commit
* `black`
    * --diff  # Shows results but does not (yet) modify the code
    * --skip-string-normalization  # Much of the code currently prefers '' over ""
* `mypy` is still heavily gated by the settings in `setup.cfg`
* `pyupgrade`

### CODE REVIEW

**Comment by RayBB** [comment]
Did you mean to add the changes to coverlib.py?

Do you want help trying to get pyupgrade working? or you're still working on it?

**Comment by cclauss** [comment]
The `coverlib.py` changes were to keep `mypy` happy.

The `pyupgrade` thing will not be needed once #5855 lands but if you can come up with an interim solution that would be warmly recieved.

**Comment by cclauss** [comment]
Did this work for you?

**Comment by RayBB** [comment]
It runs but I'm not sure if it's working as expected since it doesn't prevent me from adding a commit with a bunch of extra spaces. I believe black should be complaining about this but is not.

image

**Comment by RayBB** [comment]
We should merge #5867 before this otherwise it will block committing from gitpod. Not sure how many people commit from there but I know several people use it for testing at least.

**Comment by RayBB** [comment]
How would you feel about moving `make-lint-diff` run last since it takes the longest to run?

**Comment by cclauss** [comment]
I think that running flake8 inside the Docker container is a mistake.  Doing so takes way too long for a pre-commit job.  Why not just run it in the local directory like these other jobs.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -15,7 +15,42 @@ repos:
...
         entry: bash -c 'git diff master -U0 | docker-compose run --rm home ./scripts/flake8-diff.sh'
         language: system
+
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.0.1
+    hooks:
+    -   id: check-yaml
+    # -   id: end-of-file-fixer
+    # -   id: requirements-txt-fixer
+    # -   id: trailing-whitespace
```
Comment (RayBB) [inline .pre-commit-config.yaml:25]: Why are these commented out?

**Review by RayBB — COMMENTED** [review]
I can't speak to the type annotation changes or import changes but everything else seems good to me

File: .pre-commit-config.yaml (hunk 1 above)
Comment (cclauss) [reply .pre-commit-config.yaml:25]: Because I want to remember to add them but do not want to touch too many files in a single pull request.

**Pull request merged**